fem/ceed_algebraic.hpp

```diff
@@ -0,0 +1,169 @@
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_CEED_ALGEBRAIC_HPP
+#define MFEM_CEED_ALGEBRAIC_HPP
+
+#include "../config/config.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "fespacehierarchy.hpp"
+#include "multigrid.hpp"
+#include "libceed/ceedsolvers-utility.h"
+#include "libceed/ceed-wrappers.hpp"
+
+namespace mfem
+{
+
+/** @brief A way to use algebraic levels in a Multigrid object
+
+    This is analogous to a FiniteElementSpace but with no Mesh information,
+    constructed in a semi-algebraic way. */
+class AlgebraicCoarseSpace : public FiniteElementSpace
+{
+public:
+   AlgebraicCoarseSpace(FiniteElementSpace &fine_fes, CeedElemRestriction fine_er,
+                        int order, int dim, int order_reduction_);
+   int GetOrderReduction() const { return order_reduction; }
+   CeedElemRestriction GetCeedElemRestriction() const { return ceed_elem_restriction; }
+   CeedBasis GetCeedCoarseToFine() const { return coarse_to_fine; }
+   virtual const Operator *GetProlongationMatrix() const override { return NULL; }
+   virtual const SparseMatrix *GetRestrictionMatrix() const override { return NULL; }
+   ~AlgebraicCoarseSpace();
+
+protected:
+   int *dof_map;
+   int order_reduction;
+   CeedElemRestriction ceed_elem_restriction;
+   CeedBasis coarse_to_fine;
+};
+
+#ifdef MFEM_USE_MPI
+
+/** @brief Parallel version of AlgebraicCoarseSpace
+
+    This provides prolongation and restriction matrices for RAP-type
+    parallel operators and potential explicit assembly. */
+class ParAlgebraicCoarseSpace : public AlgebraicCoarseSpace
+{
+public:
+   ParAlgebraicCoarseSpace(
+      FiniteElementSpace &fine_fes,
+      CeedElemRestriction fine_er,
+      int order,
+      int dim,
+      int order_reduction_,
+      GroupCommunicator *gc_fine
+   );
+   virtual const Operator *GetProlongationMatrix() const override { return P; }
+   virtual const SparseMatrix *GetRestrictionMatrix() const override { return R_mat; }
+   GroupCommunicator *GetGroupCommunicator() const { return gc; }
+   HypreParMatrix *GetProlongationHypreParMatrix();
+   ~ParAlgebraicCoarseSpace();
+
+private:
+   SparseMatrix *R_mat;
+   GroupCommunicator *gc;
+   ConformingProlongationOperator *P;
+   HypreParMatrix *P_mat;
+   Array<int> ldof_group, ldof_ltdof;
+};
+
+#endif
+
+/** @brief Hierarchy of AlgebraicCoarseSpace objects for use in Multigrid object */
+class AlgebraicSpaceHierarchy : public FiniteElementSpaceHierarchy
+{
+public:
+   /** @brief Construct hierarchy based on finest FiniteElementSpace
+
+       The given space is a real (geometric) space, but the coarse spaces
+       are constructed semi-algebraically with no mesh information. */
+   AlgebraicSpaceHierarchy(FiniteElementSpace &fespace);
+   AlgebraicCoarseSpace& GetAlgebraicCoarseSpace(int level)
+   {
+      MFEM_ASSERT(level < GetNumLevels() - 1, "");
+      return static_cast<AlgebraicCoarseSpace&>(*fespaces[level]);
+   }
+   ~AlgebraicSpaceHierarchy()
+   {
+      CeedElemRestrictionDestroy(&fine_er);
```

Comment:
I can't really use the existing `InitCeedBasis` type functions, because they all assume the `FiniteElementSpace` has a `Mesh` but our algebraic spaces do not have a mesh. We could manually insert the basis/restrictions we build in `ceed_basis_map` and (maybe?) use the existing `RemoveCeedBasisAndRestriction` to delete them, but I see that as kind of a hack. I am open to cleaner ways of handling these pointers but we are interfacing with C code so I think some amount of manual creation/deletion is inevitable.

Reply:
As far as I know, the original purpose of these maps was to store the `CeedBasis` and `CeedRestriction` to avoid duplicating them in the case of multigrid, which seems to be exactly this case... From what I see, the `Mesh` is only used to access the dimension of the `FiniteElementSpace`, how do you access the dimension if you don't have a `Mesh`?

Reply:
@jeremylt Am I mistaken? My impression was that the purpose of your `CeedBasis` and `CeedElemRestriction` maps was targeting this specific case.

Reply:
Yeah, the idea behind the maps is to prevent duplication of `CeedBasis` and `CeedElemRestriction`s. Especially the non-strided restrictions, because the indices array on the fine grid be a very large amount of data and extra copies can fill up the GPU.

Reply:
I got the confirmation from @jeremylt that the `CeedElemRestriction` map was specifically designed to prevent duplicating the fine grid restrictions.

Reply:
What exactly are you asking me to do? Write my own version of `InitCeedBasisAndRestriction` that does not require a full `FiniteElementSpace`, or rewrite the existing version, or manually insert the basis and restriction we use in the map?

Reply:
I may have misunderstood you. You are only concerned about the finest `ElemRestriction`? For that one I can possibly use the existing routines.

Reply:
You were mentioning that these functions need a `Mesh` to access the dimension which is true, but if we could access the dimension of the `FiniteElementSpace` that would be sufficient and these functions wouldn't need a `Mesh`. Unfortunately, there doesn't seem to be a way to access the dimension of a `FiniteElementSpace` without the `Mesh`, wouldn't it make sense to create an empty `Mesh` just to have the dimension and the space dimension defined?